### PR TITLE
convert to state_id's in decay

### PR DESCRIPTION
**Pull request by crbates**
#664 reopened from my repo

### CODE REVIEW

**Comment by scopatz** [comment]
Hi Cameron, what happens if a state id cannot be converted to a metastable id? It seems like this could come up...

**Comment by crbates** [comment]
Right now it would just throw an error. For 99.9% of cases the half-lives are short so they can probably just be ignored, but I'm not sure what to do about this as there are ~45 states that are not listed as metastable with reasonably long half-lives based on cutting down the level list data like so:
```python
from pyne import nuc_data
import tables as tb
with tb.open_file(nuc_data) as f:
    ll = f.root.decay.level_list
    unstable = ll.read_where('(nuc_id%10000 != 0) & (nuc_id !=0) & (metastable == 0) & (half_life > 1)')
```

**Comment by scopatz** [comment]
For reference, here are the species:

```
Co68-S1 1.6
Co68-S1 1.6
Cu68-S3 225.0
Br72-S1 10.6
Br72-S1 10.6
Nb87-S1 156.0
Nb94-S1 375.78
Nb94-S1 375.78
Nb94-S1 375.78
Tc94-S1 3120.0
Tc94-S1 3120.0
Rh94-S6 25.8
Rh94-S6 25.8
Nb95-S1 311904.0
Nb95-S1 311904.0
Nb95-S1 311904.0
Tc95-S1 5270400.0
Tc95-S1 5270400.0
Tc95-S1 5270400.0
Rh95-S1 117.6
Rh95-S1 117.6
Rh95-S1 117.6
Pd95-S5 13.3
Pd95-S5 13.3
Pd95-S5 13.3
Pd95-S5 13.3
Y96-S5 9.6
Y96-S5 9.6
Tc96-S1 3090.0
Tc96-S1 3090.0
Tc96-S1 3090.0
Rh96-S1 90.6
Rh96-S1 90.6
Rh96-S1 90.6
Ag96-S1 6.9
Ag96-S1 6.9
Ag96-S1 6.9
In98-S1 1.2
Rh101-S1 374976.0
Rh101-S1 374976.0
Rh101-S1 374976.0
Rh108-S4 360.0
Rh108-S4 360.0
Ag108-S2 13822228800.0
Ag108-S2 13822228800.0
Ag108-S2 13822228800.0
In108-S1 2376.0
In108-S1 2376.0
Cd111-S3 2910.0
Cd111-S3 2910.0
In116-S1 3257.4
In116-S1 3257.4
In116-S4 2.18
In116-S4 2.18
Ag117-S1 5.34
Ag117-S1 5.34
Ag117-S1 5.34
Cd117-S2 12096.0
Cd117-S2 12096.0
In117-S1 6972.0
In117-S1 6972.0
In117-S1 6972.0
Sn117-S2 1209600.0
Sn117-S2 1209600.0
Cs117-S2 6.5
La124-S2 29.21
La124-S2 29.21
La126-S1 50.0
Sm133-S13 3.5
Cs134-S3 10483.2
Cs134-S3 10483.2
Pr134-S17 660.0
Pr134-S17 660.0
Pm138-S1 194.4
Ho154-S2 186.0
Ho154-S2 186.0
Ho154-S2 186.0
Tm154-S2 3.3
Tm154-S2 3.3
Tm154-S2 3.3
Lu154-S1 1.12
Lu154-S1 1.12
Lu157-S1 4.79
Lu157-S1 4.79
Lu157-S1 4.79
Tm158-S2 20.0
Ta160-S1 1.7
Ta161-S1 3.08
Pb185-S1 4.3
Pb185-S1 4.3
Au200-S11 67320.0
Au200-S11 67320.0
Au200-S11 67320.0
Bi216-S1 396.0
Md249-S1 1.9
```

**Comment by scopatz** [comment]
Now some of these clearly look like, ENSDF errors, for example, Sn117-S2 whose half life here is 14.0 days and whose half-life on KAERI (where it is metastable) is 13.6 days.  Similarly, Ag108-S2 is here listed with a half-life of 438 years but on KAERI is 418 years.  I think the other long-lived species here will also end up being mistakes. Of the rest, they really are not metastable.

So how do we go about confirming and reporting these errors?

**Comment by crbates** [comment]
> Hi Cameron, what happens if a state id cannot be converted to a metastable id? It seems like this could come up...

This currently can't happen because all_nucs is limited to the list of metastable and stable nuclides. This PR already does what we discussed in the dev call today ie. all non-metastable states are zeroed out. The way that children are currently calculated it's actually not possible to determine state population at all so everything is just assumed to go to the ground state. In order to calculate state populations one would need to build a decay scheme which is still on my to do list.

**Comment by crbates** [comment]
ping @scopatz. The only test this fails is the existing one in `material`. This should be fixed by updating `decay.cpp` on the server.

File: src/decaygen.py [hunk 1]
```diff
@@ -82,9 +83,16 @@
   double out [{{ nucs|length }}] = {};  // init to zero
   map<int, double> outcomp;
   
-  // body
+  //convert to state id's
+  std::map<int, double> scomp;
   map<int, double>::const_iterator it = comp.begin();
   for (; it != comp.end(); ++it) {
+      scomp.insert(std::pair<int,double>(nucname::id_to_state_id(it->first),it->second));
```
Comment (scopatz) [inline src/decaygen.py]: SG: space after comma.

**Comment by scopatz** [comment]
Looking at this again, wouldn't it be faster and less memory intensive to change line 117 of decaygen.py to the following:

```diff
-   switch (it->first) {
+   switch (nucname::id_to_state_id(it->first)) {
```

**Comment by scopatz** [comment]
Also, the `decay_data_children()` function could use some docs, but maybe that should be a separate issue :).  I didn't realize that it took/returned state ids.

**Comment by crbates** [comment]
>Looking at this again, wouldn't it be faster and less memory intensive to change line 117 of decaygen.py to the following:

>```diff
-   switch (it->first) {
+   switch (nucname::id_to_state_id(it->first)) {
```

Yeah, I was originally thinking about having the option of using state_id's in the mat if desired which is why I used the pattern in this PR but I think for now this would be simpler/make more sense.

**Comment by crbates** [comment]
> yeah I forgot we didn't use `automodule` for decay. The docstring does specify `state_id` for `decay_data_children`. I'm still not sure whether the decay dataset data is a better source or the level data is for children and half lives. I switched it to using decay dataset data because of a problem with the level data entries which I think was resolved by #671.

**Comment by crbates** [comment]
ping @scopatz. I think this is the right way to go about this.

**Comment by scopatz** [comment]
Hmm, why do the tests pass now?

**Comment by scopatz** [comment]
ie I thought you had removed a specied that was known not to show up in U-235...

File: src/decaygen.py [hunk 2]
```diff
@@ -88,7 +89,7 @@
     switch (nucname::znum(it->first)) {
       {{ cases|indent(6) }}
       default:
-        outcomp.insert(*it);
+        outcomp.insert(std::make_pair(nucname::id_to_state_id(it->first), it->second));
```
Comment (scopatz) [inline src/decaygen.py]: Actually, I don't think that this is right because outcomp is keyed by id, not state id.

File: src/decaygen.py [hunk 3]
```diff
@@ -114,10 +115,10 @@
 ELEM_FUNC = ENV.from_string("""
 void decay_{{ elem|lower }}(double t, std::map<int, double>::const_iterator &it, std::map<int, double> &outcomp, double (&out)[{{ nucs|length }}]) {
   //using std::exp2;
-  switch (it->first) {
+  switch (nucname::id_to_state_id(it->first)) {
     {{ cases|indent(4) }}
     } default: {
-      outcomp.insert(*it);
+      outcomp.insert(std::make_pair(nucname::id_to_state_id(it->first), it->second));
```
Comment (scopatz) [inline src/decaygen.py]: same here,

**Comment by crbates** [comment]
Something has gone terribly wrong with my `virtualenv` setup. I will fix this once I figure out what happened.

**Comment by scopatz** [comment]
Good luck!

**Comment by scopatz** [comment]
is this ready, mi amigo?

**Comment by crbates** [comment]
ping @scopatz. This will need an updated benchmark file I think.

**Comment by scopatz** [comment]
I doubt that the benchmark file will need to be updated. It just contains the results of running origen many times.

File: pyne/data.pyx [hunk 4]
```diff
@@ -1044,6 +1044,52 @@ def decay_children(nuc, use_metastable=True):
...
+    nuc : int
+        input nuclide in state id form
+
+    Returns
+    -------
+    ids : list of ints
+        list of all known decay children from ensdf data.
+    """
+
+    return list(set(decay_children(nuc, False)) & set(decay_data_children(nuc)))
```
Comment (scopatz) [inline pyne/data.pyx]: Does this need to return a list?  A set seems more approriate.

Comment (scopatz) [inline pyne/data.pyx]: Also, shouldn't this be `set(decay_children(nuc, False)) | set(decay_data_children(nuc))`?  That is, don't we want the union and not the intersection?

Comment (crbates) [inline pyne/data.pyx]: oops. thats why `decay.cpp` was smaller. Now it's close to 19MB!

**Comment by crbates** [comment]
@scopatz I'm not sure whats going on here. for some reason bi-209's case now looks like:
```c++
    case 832090000: {
      out[3260] += it->second;
      break;
    }
```

**Comment by scopatz** [comment]
That means it is being iterpreted as stable.

**Comment by crbates** [comment]
ping @scopatz. This still has failures on the full suite with the new `decay.cpp` (and it's 18.6 MB!). How do I go about resolving decay suite issues?

**Comment by scopatz** [comment]
You can add species to the METASTABLE_BLACKLIST if you know that the comparison is non-sensocal

**Comment by crbates** [comment]
ok, the black list has been updated (It's SHORTER by 5!) this passes for me locally.

**Comment by scopatz** [comment]
In the immemorial words of Street Sharks, 

jawsome_fullpic_artwork

Do you want to do the honors of uploading the new decay code or should I?

**Comment by crbates** [comment]
I'm on it.

**Comment by scopatz** [comment]
thx

**Pull request merged**